src/utilities/codemirror/vml.ts

```diff
@@ -0,0 +1,199 @@
+  },
+  name: 'vml',
+  parser: parser.configure({
+    props: [
+      foldNodeProp.add(FoldBehavior),
+      styleTags({
+        ADD: t.arithmeticOperator,
+        ASSIGNMENT: t.updateOperator,
+        BLOCK: t.namespace,
+        BODY: t.namespace,
+        Comment: t.comment,
+        DAY_TIME_CONST: t.className,
+        DECLARE: t.keyword,
+        DELAY_BY: t.keyword,
+        DIVIDE: t.arithmeticOperator,
+        DOUBLE_CONST: t.number,
+        END_BODY: t.namespace,
+        END_MODULE: t.namespace,
+        EXTERNAL_CALL: t.controlKeyword,
+        FULL_TIME_CONST: t.className,
+        HEX_CONST: t.number,
+        INPUT: t.keyword,
+        INT_CONST: t.number,
+        ISSUE: t.controlKeyword,
+        MODULE: t.namespace,
+        MODULO: t.arithmeticOperator,
+        MULTIPLY: t.arithmeticOperator,
+        POWER: t.arithmeticOperator,
+        RETURN: t.keyword,
+        SEQUENCE: t.macroName,
+        SHORT_TIME_CONST: t.className,
+        SPACECRAFT_TIME_CONST: t.className,
+        SPAWN: t.controlKeyword,
+        STRING_CONST: t.string,
+        SUBTRACT: t.arithmeticOperator,
+        TIMEOUT: t.keyword,
+        UINT_CONST: t.number,
+        VML_EOF: t.docComment,
+        VML_HEADER: t.docComment,
+        Variable_name: t.variableName,
+        WAIT: t.keyword,
+        ...Object.fromEntries(
+          [
+            TOKEN_EXTERNAL_CALL,
+            TOKEN_CALL,
+            TOKEN_SPAWN,
+            TOKEN_ISSUE,
+            // conditionals
+            TOKEN_IF,
+            TOKEN_THEN,
+            TOKEN_ELSE_IF,
+            TOKEN_ELSE,
+            TOKEN_END_IF,
+            // for
+            TOKEN_FOR,
+            TOKEN_TO,
+            TOKEN_STEP,
+            TOKEN_DO,
+            TOKEN_END_FOR,
+            // while
+            TOKEN_WHILE,
+            TOKEN_END_WHILE,
+          ].map(token => [token, t.controlKeyword]),
+        ),
+      }),
+    ],
+  }),
+});
+
+export function setupVmlLanguageSupport(
+  autocomplete?: (context: CompletionContext) => CompletionResult | null,
+): LanguageSupport {
+  if (autocomplete) {
+    const autocompleteExtension = VmlLanguage.data.of({ autocomplete });
+    return new LanguageSupport(VmlLanguage, [vmlBlockFolder, autocompleteExtension]);
+  } else {
+    return new LanguageSupport(VmlLanguage, [vmlBlockFolder]);
+  }
+}
+
+export function vmlHighlightBlock(viewUpdate: ViewUpdate): SyntaxNode[] {
+  const tree = syntaxTree(viewUpdate.state);
+  const selectionLine = viewUpdate.state.doc.lineAt(viewUpdate.state.selection.asSingle().main.from);
+  const leadingWhiteSpaceLength = selectionLine.text.length - selectionLine.text.trimStart().length;
+  const updatedSelectionNode = tree.resolveInner(selectionLine.from + leadingWhiteSpaceLength, 1);
+  // walk up the tree to Time_tagged_statement, and then back down to the block command e.g. "ELSE"
+  const timeTaggedNode = getNearestAncestorNodeOfType(updatedSelectionNode, [RULE_TIME_TAGGED_STATEMENT]);
```

Comment:
Same question did we want to move these over to the editor or a separate file?

Reply:
I'd prefer to keep this out of the editor, hopefully we move this to an optional plugin